snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,206 @@
+
+    g_autoptr(GError) error = NULL;
+    g_autoptr(GPtrArray) notices = snapd_client_get_notices_finish(source, res, &error);
+
+    if (error != NULL) {
+        if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+            g_signal_emit_by_name(self, "error-event", error);
+        self->running = FALSE;
+        g_clear_object(&self->cancellable);
+        self->cancellable = g_cancellable_new();
+        g_object_unref(self);
+        return;
+    }
+
+    if (notices != NULL) {
+        gboolean first_run = self->last_notice == NULL;
+        for (int i = 0; i < notices->len; i++) {
+            g_autoptr(SnapdNotice) notice = g_object_ref(notices->pdata[i]);
+
+            if (self->last_notice == NULL || snapd_notice_compare_last_occurred(self->last_notice, notice) <= 0) {
+                g_clear_object(&self->last_notice);
+                self->last_notice = g_object_ref(notice);
+            }
+            g_signal_emit_by_name(self, "notice-event", notice, first_run);
+        }
+    }
+    begin_monitor(self); // continue waiting for new notifications
+}
+
+static void begin_monitor (SnapdNoticesMonitor *self) {
+    snapd_client_notices_set_after_notice (self->client, self->last_notice);
+    g_autoptr(GDateTime) last_date_time = self->last_notice == NULL ? NULL : (GDateTime *) snapd_notice_get_last_occurred (self->last_notice);
+    snapd_client_get_notices_async(self->client,
+                                   last_date_time,
+                                   2000000000000000, // "infinity" (there is a limit in snapd, around 9 billion seconds)
+                                   self->cancellable,
+                                   (GAsyncReadyCallback)monitor_cb,
+                                   self);
+}
+
+/**
+ * snapd_notices_monitor_start:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Starts the asynchronous listening proccess, that will wait for new
+ * notices and emit a "notice-event" signal with the new notice as
+ * parameter.
+ *
+ * Returns: FALSE if there was an error, TRUE if everything worked fine
+ * and the object is listening for events.
+ */
+gboolean snapd_notices_monitor_start (SnapdNoticesMonitor *self, GError **error) {
+
+    g_return_val_if_fail((error == NULL) || (*error == NULL), FALSE);
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), FALSE);
+    if (self->running) {
+        *error = g_error_new(SNAPD_ERROR, SNAPD_ERROR_ALREADY_RUNNING, "The notices monitor is already running.");
+        return FALSE;
+    }
+    self->running = TRUE;
+    begin_monitor(g_object_ref(self));
+    return TRUE;
+}
+
+/**
+ * snapd_notices_monitor_stop:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Stops the asynchronous listening proccess started with #snapd_notices_monitor_start.
+ *
+ * Returns: FALSE if there was an error, TRUE if everything worked fine.
+ */
+gboolean snapd_notices_monitor_stop (SnapdNoticesMonitor *self, GError **error) {
+
```

Comment:
whitespace

Reply:
Ops... also fixed the style to make it consistent with the rest of the code.